### PR TITLE
Global Error Handling

### PR DESCRIPTION
**Pull request by cfryerdev**
**Note: Visual studio updated some namespaces in the feature tests automatically, they are included in this PR to true up those files.**

This adds global error handling which will return either the resource you requested with an expected response code... OR, a standardized error payload that does not change. This handles model state validation as well as thrown and uncaught exceptions. 

This also allows you to throw exceptions with specific types, override error codes and http status codes and is rather extensible for error codes using enums or a resource lookup. I added an example to the movie controller that you can test in swagger. This is not meant to be a "one size fits all" solution and should be extended when client requirements dictate different error code lookups. I generally also add a configuration setting in appsettings to hide/show the stacktrace in the payload, should you want to expose that in lower environments.

This is also incredibly important for modern web projects that consume a blackslope api, as it allows interceptors to be added to http/fetch functionality to automatically trigger should it see an error and pop a toast, as the payload is always consistent and full of details.

This has been added to numerous slalom projects to great success, so finally taking the time to add it to the root project.  Please reach out to me on slack should you have any questions or want to see it in action.

### CODE REVIEW

**Comment by cfryerdev** [comment]
image

**Comment by cfryerdev** [comment]
image

File: src/BlackSlope.Api/Extensions/MvcExtensions.cs [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+namespace BlackSlope.Api.Extensions
+{
+    public static class MvcExtensions
+    {
+        public static IServiceCollection AddMvcApi(this IServiceCollection services)
+        {
+            services
+                .AddMvc(mvcOptions =>
+                {
+                    mvcOptions.Filters.Add(new ModelstateValidationFilter());
```
Comment (theczorn) [inline src/BlackSlope.Api/Extensions/MvcExtensions.cs:13]: sp: `Modelstate` to `ModelState`

File: src/BlackSlope.Api.Common/Exceptions/ExceptionListItem.cs [hunk 2]
```diff
@@ -0,0 +1,15 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace BlackSlope.Api.Common.Exceptions
+{
+    public class ExceptionListItem
```
Comment (theczorn) [inline src/BlackSlope.Api.Common/Exceptions/ExceptionListItem.cs:9]: Nit: `HandledSuggestionListItem`?

File: src/BlackSlope.Api.Common/Exceptions/ExceptionType.cs [hunk 3]
```diff
@@ -0,0 +1,16 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Text;
+
+namespace BlackSlope.Api.Common.Exceptions
+{
+    public enum ExceptionType
```
Comment (theczorn) [inline src/BlackSlope.Api.Common/Exceptions/ExceptionType.cs:7]: Nit: `HandledExceptionType`?

File: src/BlackSlope.Api.Common/Exceptions/HandledException.cs [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+        }
+
+        /// <summary>
+        /// Gets the exception list in a readable form for restful consumption.
+        /// </summary>
+        /// <returns></returns>
+        public List<ExceptionListItem> GetExceptionList()
+        {
+            var collection = new List<ExceptionListItem>();
+            if (InnerExceptions != null && InnerExceptions.Any())
```
Comment (theczorn) [inline src/BlackSlope.Api.Common/Exceptions/HandledException.cs:85]: Can be streamlined with `InnerExceptions?.Any()` I believe...

**Review by theczorn — APPROVED** [review]
Minor syntax quirks, lgtm.

File: src/BlackSlope.Api.Common/Exceptions/HandledResult.cs [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+using System.Collections.Generic;
+using System.Net;
+
+namespace BlackSlope.Api.Common.Exceptions
+{
+    public class HandledResult<T> where T : Exception
+    {
+        private const string BadRequestDescriptor = "Bad Request";
+
+        private const string ServerErorDescriptor = "Internal Server Error";
```
[the remark this answers is not in the document]
Comment (cfryerdev) [reply src/BlackSlope.Api.Common/Exceptions/HandledResult.cs:11]: Will fix, thanks.

File: src/BlackSlope.Api.Common/Exceptions/ExceptionListItem.cs (hunk 2 above)
Comment (cfryerdev) [reply src/BlackSlope.Api.Common/Exceptions/ExceptionListItem.cs:9]: I prefer the current name, but thank you for the suggestion

File: src/BlackSlope.Api.Common/Exceptions/HandledException.cs (hunk 4 above)
Comment (cfryerdev) [reply src/BlackSlope.Api.Common/Exceptions/HandledException.cs:85]: Error: Cannot implicitly convert type 'bool?' to 'bool'. Leaving for now.

File: src/BlackSlope.Api/Extensions/MvcExtensions.cs (hunk 1 above)
Comment (cfryerdev) [reply src/BlackSlope.Api/Extensions/MvcExtensions.cs]: Renamed

File: src/BlackSlope.Api.Common/Exceptions/ExceptionType.cs (hunk 3 above)
Comment (cfryerdev) [reply src/BlackSlope.Api.Common/Exceptions/ExceptionType.cs:7]: Ill leave it for now, on the fence.

**Pull request merged**